2021-Firmware-Bootcamp/Core/Src/main.c

```diff
@@ -0,0 +1,208 @@
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  // Defining variables, array of 3 because sending 3 bytes
+  uint8_t tx_data[3] = {0x01, 0x80, 0x00};
+  uint8_t rx_data[3];
+  const uint16_t LENGTH = 3;
+  const uint32_t TIMEOUT = 100;
+  const uint8_t RESOLUTION = 10;
+  const uint16_t FIVE_PERCENT_DUTY_CYCLE = 1500;
+  // Needs to store 10 bits
+  uint16_t data;
+  // Start Timer
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+	  // Need to pull CS LOW to active SPI PB8
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+	  // Transmit and receive from SPI and see if it is OK
+	  HAL_StatusTypeDef hal_status = HAL_SPI_TransmitReceive(&hspi1, tx_data, rx_data, LENGTH, TIMEOUT);
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+	  if (hal_status == HAL_OK){
+		  // Set the count value got from ADC, the last 10 bits
+		  data = (rx_data[2] | (rx_data[1] & 0x03)) / (1 << RESOLUTION) * FIVE_PERCENT_DUTY_CYCLE + FIVE_PERCENT_DUTY_CYCLE ;
```

Comment:
so rx_data[2] and rx_data[1] make up 10 bits but you are shoving them all into 8 bits, you need to shift over rx_data[1] 8 bits to the left before ORing with rx_data[2]